plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -307,4 +330,20 @@ public int read(byte[] b, int off, int len) throws IOException {
             }
         };
     }
+
+    @Override
+    public Map<String, Long> stats() {
+        return stats.toMap();
+    }
+
+    private static class Stats {
```

Comment:
I'm thinking about moving this class (that's shared across the 3 repositories) into `org.elasticsearch.common.blobstore`. Additionally I would use an Enum instead of using a String to represent the operation type. wdyt? We could get rid of some duplication.

Reply:
I'm not so sure about that. The classes are the same now because we only track `GET` and `LIST` which are the same across all blob store in how they work and how we want to track them.
But once we add writes, we'll have to track different kinds of requests (e.g. for GCS we'll track the last request in a resumable upload while in S3 we'd track the requests for starting and ending the a multi-part upload as well as the `PUT`s for the individual parts).
If we just have one class and an `Enum` we're either pretty limited by the resolution on the request type we can express or we'll have some merged `Enum` that contains the implementation details of the various request types for all the plugins.

I guess we could make the thing a little more generic if we wanted to though? We could have an Enum per plugin for the request types and then do something generic like this:

```java
public class BlobStoreStats <T> {

    private final Map<T, AtomicLong> counts = ConcurrentCollections.newConcurrentMap();

    public void track(T opType) {
        counts.computeIfAbsent(opType, k -> new AtomicLong(0)).incrementAndGet();
    }
}
```

and then use it like e.g.

```java
  stats.track(S3Op.GET);
```

How about that?

Reply:
I agree, that approach looks more flexible than what I was proposing. I was just trying to keep things simple as my initial thought was that we would classify the same operations across different implementations. I'll tackle that in a different PR.